python/openai/openai_frontend/frontend/fastapi/middleware/api_restriction.py

```diff
@@ -0,0 +1,224 @@
+            app: The FastAPI application instance
+            restricted_apis: RestrictedFeatures instance containing the restriction configuration
+        """
+        super().__init__(app)
+        self.restricted_apis = restricted_apis
+
+    def _get_auth_header(self, request: Request) -> tuple[str, str] | None:
+        request_method = request.method
+        request_path = request.url.path
+
+        # Check each restricted API to see if the request matches
+        for (
+            restricted_api,
+            auth_spec,
+        ) in self.restricted_apis.RestrictionDict().items():
+            # Check each endpoint in the API
+            for restricted_endpoint in ENDPOINT_MAPPING[restricted_api]:
+                restricted_method, restricted_path = restricted_endpoint.split(" ")
+
+                # Match both HTTP method and path prefix
+                if request_method == restricted_method and request_path.startswith(
+                    restricted_path
+                ):
+                    return auth_spec
+        return None
+
+    async def dispatch(self, request: Request, call_next):
+        """
+        Main middleware dispatch method that processes each incoming request.
+
+        Args:
+            request: The incoming HTTP request
+            call_next: The next middleware/handler in the chain
+
+        Returns:
+            Response: Either the next handler's response or a 401 authentication error
+        """
+        # Check if the request matches any restricted patterns
+        auth_header = self._get_auth_header(request)
+
+        # If request not restricted, proceed with the request
+        if not auth_header:
+            return await call_next(request)
+
+        # Check authentication for the matching restricted endpoint
+        auth_result = self._check_authentication(request, auth_header)
+        if auth_result["valid"]:
+            # Authentication passed, allow request to proceed
+            return await call_next(request)
+        else:
+            # Authentication failed, return 401 error
+            return JSONResponse(
+                status_code=401,
+                content={
+                    "error": {
+                        "message": auth_result["message"],
+                        "type": "authentication_error",
+                        "code": "invalid_auth",
+                    }
+                },
+            )
+
+    def _check_authentication(self, request: Request, auth_header: tuple[str, str]):
+        """
+        Check if the request contains valid authentication headers.
+
+        Args:
+            request: The incoming HTTP request
+            auth_header: Tuple of (expected_header_key, expected_header_value)
+
+        Returns:
+            dict: {"valid": bool, "message": str} - Authentication result and error message if invalid
+        """
+        expected_key, expected_value = auth_header
+        print(f"expected_key: {expected_key}")
```

Comment:
I think we would leave this out until we get logging levels - @rmccorm4 , @whoisj  what do you think?